File: P03Heat/exts/Algos.c
#include "Exts.h"

// Calculate the residue vector
void calc_res(int size, sparse_matrix_t mat, double* vec, double* sol, double* res)
{
    int n = size;
    double* b = vec, * x = sol;
    sparse_matrix_t a = mat;
    double* r = res;
    
    cblas_dcopy(n, b, 1, r, 1);
    mkl_sparse_d_mv(SPARSE_OPERATION_NON_TRANSPOSE, -1.0, a, (struct matrix_descr){SPARSE_MATRIX_TYPE_GENERAL, 0, 0}, x, 1.0, r);
    
    return ;
}

// Calculate the squared norm of residue
double calc_res_norm2(int size, sparse_matrix_t mat, double* vec, double* sol, double* work)
{
    int n = size;
    double* b = vec, * x = sol;
    sparse_matrix_t a = mat;
    double* r = work;

    calc_res(n, a, b, x, r);
    double rho = cblas_ddot(n, r, 1, r, 1);
    
    return rho;
}

// Calculate the lower part of a banded matrix
void calc_lower_band(int size, double* data, int* ind, int* ptr, double* band, int width)
{
    int n = size, w = width;

    for (int i = 0; i < n * (w+1); i++)
        band[i] = 0.0;

    int pos = *ptr;
    ptr++;
    for (int i = 0; i < n; i++)
    {
        for (; pos < *ptr; pos++)
            if (ind[pos] > i)
                break;
            else
                band[(w+1) * ind[pos] + i - ind[pos]] = data[pos];
        pos = *ptr;
        ptr++;
    }

    return ;
}

// Solve a system using Cholesky decomposition
int evolve_cho(int size, double* data, int* ind, int* ptr, double* init, int width, int rep)
{
    int n = size, w = width;
    sparse_matrix_t a;
    mkl_sparse_d_create_csr(&a, SPARSE_INDEX_BASE_ZERO, n, n, ptr, ptr+1, ind, data);
    double* u = init;
    double* l = malloc((width+1) * n * sizeof(double));
    
    calc_lower_band(n, data, ind, ptr, l, w);

    LAPACKE_dpbtrf(LAPACK_COL_MAJOR, 'L', n, w, l, w+1);

    int ctr = 0;
    for (int i = 0; i < rep; i++)
    {
        cblas_dtbsv(CblasColMajor, CblasLower, CblasNoTrans, CblasNonUnit, n, w, l, w+1, u, 1);
        cblas_dtbsv(CblasColMajor, CblasLower, CblasTrans, CblasNonUnit, n, w, l, w+1, u, 1);

        ctr++;
    }

    free(l);
    mkl_sparse_destroy(a);

    return ctr;
}

// Perform a Gauss-Seidel iteration step
void step_gs(int size, sparse_matrix_t mat, double* vec, double* init)
{
    int n = size;
    double* data;
    int* ind, * ptrb, * ptre;
    int temp;
    mkl_sparse_d_export_csr(mat, &temp, &temp, &temp, &ptrb, &ptre, &ind, &data);

    for (int i = 0; i < n; i++)
    {
        double acc = 0.0, coe;
        for (int pos = *ptrb; pos < *ptre; pos++)
            if (ind[pos] != i)
                acc += data[pos] * init[ind[pos]];
            else
                coe = data[pos];
        init[i] = (vec[i] - acc) / coe;
        ptrb++, ptre++;
    }
    
    return ;
}

// Solve a system using Gauss-Seidel
int solve_gs(int size, sparse_matrix_t mat, double* vec, double* init, double eps, int chk, double* work)
{
    int n = size;
    double* b = vec, * x = init;
    sparse_matrix_t a = mat;
    double* r = work;

    double rho_0 = calc_res_norm2(n, a, b, x, r);

    int ctr = 0;
    while (1)
    {
        step_gs(n, a, b, x);

        ctr++;
        if (ctr % chk == 0)
        {
            double rho = calc_res_norm2(n, a, b, x, r);
            if (rho < rho_0 * eps * eps)
                break;
        }
    }

    return ctr;
}

// Evolve a PDE using Gauss-Seidel
int evolve_gs(int size, double* data, int* ind, int* ptr, double* init, double eps, int chk, int rep)
{
    int n = size;
    sparse_matrix_t a;
    mkl_sparse_d_create_csr(&a, SPARSE_INDEX_BASE_ZERO, n, n, ptr, ptr+1, ind, data);
    double* u = init;
    double* w = malloc(2*n * sizeof(double));
    double* t = w + n;

    int ctr = 0;
    for (int i = 0; i < rep; i++)
    {
        cblas_dcopy(n, u, 1, t, 1);
        ctr += solve_gs(n, a, t, u, eps, chk, w);
    }

    free(w);
    mkl_sparse_destroy(a);

    return ctr;
}

// Solve a system using steepest descent
int solve_sd(int size, sparse_matrix_t mat, double* vec, double* init, double eps, double* work)
{
    int n = size;
    double* b = vec, * x = init;
    sparse_matrix_t a = mat;
    double* r = work, * t = work + n;
    
    double rho = calc_res_norm2(n, a, b, x, r);
    double rho_0 = rho;
    
    int ctr = 0;
    while (1)
    {
        mkl_sparse_d_mv(SPARSE_OPERATION_NON_TRANSPOSE, 1.0, a, (struct matrix_descr){SPARSE_MATRIX_TYPE_GENERAL, 0, 0}, r, 0.0, t);
        double alpha = rho / cblas_ddot(n, r, 1, t, 1);
        cblas_daxpy(n, alpha, r, 1, x, 1);
        cblas_daxpy(n, -alpha, t, 1, r, 1);
        rho = cblas_ddot(n, r, 1, r, 1);
        
        ctr++;
        if (rho < rho_0 * eps * eps)
            break;
    }

    return ctr;
}

// Evolve a PDE using steepest descent
int evolve_sd(int size, double* data, int* ind, int* ptr, double* init, double eps, int rep)
{
    int n = size;
    sparse_matrix_t a;
    mkl_sparse_d_create_csr(&a, SPARSE_INDEX_BASE_ZERO, n, n, ptr, ptr+1, ind, data);
    double* u = init;
    double* w = malloc(3*n * sizeof(double));
    double* t = w + (2*n);

    int ctr = 0;
    for (int i = 0; i < rep; i++)
    {
        cblas_dcopy(n, u, 1, t, 1);
        ctr += solve_sd(n, a, t, u, eps, w);
    }

    free(w);
    mkl_sparse_destroy(a);

    return ctr;
}

// Solve a system using conjugate descent
int solve_cg(int size, sparse_matrix_t mat, double* vec, double* init, double eps, double* work)
{
    int n = size;
    double* b = vec, * x = init;
    sparse_matrix_t a = mat;
    double* r = work, * p = work + n, * t = work + (2*n);
    
    double rho = calc_res_norm2(n, a, b, x, r);
    double rho_0 = rho, rho_old;
    
    int ctr = 0;
    while (1)
    {
        if (!ctr)
            cblas_dcopy(n, r, 1, p, 1);
        else
        {
            double beta = rho / rho_old;
            cblas_dscal(n, beta, p, 1);
            cblas_daxpy(n, 1.0, r, 1, p, 1);
        }

        mkl_sparse_d_mv(SPARSE_OPERATION_NON_TRANSPOSE, 1.0, a, (struct matrix_descr){SPARSE_MATRIX_TYPE_GENERAL, 0, 0}, p, 0.0, t);

        double alpha = rho / cblas_ddot(n, p, 1, t, 1);
        cblas_daxpy(n, alpha, p, 1, x, 1);
        cblas_daxpy(n, -alpha, t, 1, r, 1);

        rho_old = rho;
        rho = cblas_ddot(n, r, 1, r, 1);
        
        ctr++;
        if (rho < rho_0 * eps * eps)
            break;
    }

    return ctr;
}

// Evolve a PDE using conjugate gradient
int evolve_cg(int size, double* data, int* ind, int* ptr, double* init, double eps, int rep)
{
    int n = size;
    sparse_matrix_t a;
    mkl_sparse_d_create_csr(&a, SPARSE_INDEX_BASE_ZERO, n, n, ptr, ptr+1, ind, data);
    double* u = init;
    double* w = malloc(4*n * sizeof(double));
    double* t = w + (3*n);

    int ctr = 0;
    for (int i = 0; i < rep; i++)
    {
        cblas_dcopy(n, u, 1, t, 1);
        ctr += solve_cg(n, a, t, u, eps, w);
    }

    free(w);
    mkl_sparse_destroy(a);

    return ctr;
}

// Perform a multigrid step
void step_mg(int size, sparse_matrix_t* mats, sparse_matrix_t* ints, double* vec, double* init, double* work, int* num, int depth)
{
    int n = size;
    sparse_matrix_t* as = mats, * is = ints;
    double* b = vec, * x = init;
    double* r = work;
    int s = (n-1) * (n-1);

    for (int i = 0; i < *num; i++)
        step_gs(s, as[0], b, x);
    
    if (depth > 1)
    {
        cblas_dcopy(s, b, 1, r, 1);
        mkl_sparse_d_mv(SPARSE_OPERATION_NON_TRANSPOSE, -1.0, as[0], (struct matrix_descr){SPARSE_MATRIX_TYPE_GENERAL, 0, 0}, x, 1.0, r);
        mkl_sparse_d_mv(SPARSE_OPERATION_TRANSPOSE, 1.0, is[0], (struct matrix_descr){SPARSE_MATRIX_TYPE_GENERAL, 0, 0}, r, 0.0, b + s);
        for (int i = 0; i < (n/2-1) * (n/2-1); i++)
            (x + s)[i] = 0.0;
        step_mg(n/2, as+1, is+1, b + s, x + s, r + s, num+1, depth-1);
        mkl_sparse_d_mv(SPARSE_OPERATION_NON_TRANSPOSE, 1.0, is[0], (struct matrix_descr){SPARSE_MATRIX_TYPE_GENERAL, 0, 0}, x + s, 1.0, x);
    }

    return ;
}

// Solve a system using multigrid
int solve_mg(int size, sparse_matrix_t* mats, sparse_matrix_t* ints, double* vec, double* init, double* work, int* num, int depth, double eps, int chk)
{
    int n = size;
    sparse_matrix_t* as = mats, * is = ints;
    double* b = vec, * x = init;
    double* r = work;
    int s = (n-1) * (n-1);
    
    double rho_0 = calc_res_norm2(s, as[0], b, x, r);

    int ctr = 0;
    while (1)
    {
        step_mg(n, as, is, b, x, r, num, depth);

        ctr++;
        if (ctr % chk == 0)
        {
            double rho = calc_res_norm2(s, as[0], b, x, r);
            if (rho < rho_0 * eps * eps)
                break;
        }
    }

    return ctr;
}

// Evolve a PDE using multigrid
int evolve_mg(int size, double** datas, int** inds, int** ptrs, double* init, int* num, int depth, double eps, int chk, int rep)
{
    int n = size;
    double* x = init;
    sparse_matrix_t* as = malloc(depth * sizeof(sparse_matrix_t)), * is = malloc((depth-1) * sizeof(sparse_matrix_t));
    double* bp, * xp, * rp;
    int s = (n-1) * (n-1);

    int acc = 0, temp = n;
    for (int i = 0; i < depth; i++)
    {
        acc += (temp-1) * (temp-1);
        temp /= 2;
    }
    bp = malloc(acc * sizeof(double)), xp = malloc(acc * sizeof(double)), rp = malloc(acc * sizeof(double));

    temp = n;
    mkl_sparse_d_create_csr(as, SPARSE_INDEX_BASE_ZERO, s, s, ptrs[0], ptrs[0]+1, inds[0], datas[0]);
    for (int i = 1; i < depth; i++)
    {
        mkl_sparse_d_create_csr(is + (i-1), SPARSE_INDEX_BASE_ZERO, (temp-1) * (temp-1), (temp/2-1) * (temp/2-1), ptrs[i], ptrs[i]+1, inds[i], datas[i]);
        sparse_matrix_t t;
        mkl_sparse_spmm(SPARSE_OPERATION_NON_TRANSPOSE, as[i-1], is[i-1], &t);
        mkl_sparse_spmm(SPARSE_OPERATION_TRANSPOSE, is[i-1], t, as + i);
        mkl_sparse_destroy(t);
        temp /= 2;
    }
    
    cblas_dcopy(s, x, 1, xp, 1);
    
    int ctr = 0;
    for (int i = 0; i < rep; i++)
    {
        cblas_dcopy(s, xp, 1, bp, 1);
        ctr += solve_mg(n, as, is, bp, xp, rp, num, depth, eps, chk);
    }
    
    cblas_dcopy(s, xp, 1, x, 1);

    for (int i = 1; i < depth; i++)
    {
        mkl_sparse_destroy(as[i]);
        mkl_sparse_destroy(is[i-1]);
    }
    mkl_sparse_destroy(as[0]);
    free(bp), free(xp), free(rp);
    free(as), free(is);

    return ctr;
}

File: P03Heat/algos.py
import numpy
import scipy.linalg
import scipy.sparse
import scipy.sparse.linalg
import utils
import exts


def pde_exp(size_h, size_t, init, time):
    """Evolve the PDE with explicit scheme and Python SuperLU binding"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye - dt / dx**2 * lap
    for i in range(m):
        u = a.dot(u)
    u = u.reshape((n-1, n-1))
    return u


def pde_imp(size_h, size_t, init, time):
    """Evolve the PDE with implicit scheme and Python SuperLU binding"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    a_fac = scipy.sparse.linalg.splu(a.transpose())
    for i in range(m):
        u = a_fac.solve(u)
    u = u.reshape((n-1, n-1))
    return u


def pde_cn(size_h, size_t, init, time):
    """Evolve the PDE with Crank-Nicolson scheme and Python SuperLU binding"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye - dt / dx**2 / 2.0 * lap
    b = eye + dt / dx**2 / 2.0 * lap
    b_fac = scipy.sparse.linalg.splu(b.transpose())
    for i in range(m):
        u = a.dot(u)
        u = b_fac.solve(u)
    u = u.reshape((n-1, n-1))
    return u


def pde_imp_cho_py(size_h, size_t, init, time):
    """Evolve the PDE using implicit scheme, Cholesky decomposition and Python"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    a = a.todia()
    ab = numpy.zeros((n, (n-1)**2))
    ab[0] = a.diagonal(k=0)
    ab[1, :-1] = a.diagonal(k=1)
    ab[n-1, :1-n] = a.diagonal(k=n-1)
    l = scipy.linalg.cholesky_banded(ab, lower=True)
    l_t = numpy.zeros((n, (n-1)**2))
    for i in range(n):
        l_t[n-i-1, i:] = l[i, :(n-1)**2-i]
    ctr = 0
    for i in range(m):
        u = scipy.linalg.solve_banded((n-1, 0), l, u)
        u = scipy.linalg.solve_banded((0, n-1), l_t, u)
        ctr += 1
    u = u.reshape((n-1, n-1))
    return u, ctr


def solve_gs(mat, mat_u, mat_dl_fac, vec, init, eps, chk):
    """Gauss-Seidel to solve a system"""
    b, x = vec, init
    a, a_u, a_dl_fac = mat, mat_u, mat_dl_fac
    w = a_dl_fac.solve(b)
    delta_0 = numpy.linalg.norm(b - a.dot(x), 2)
    ctr = 0
    while True:
        x = a_dl_fac.solve(a_u.dot(x)) + w
        ctr += 1
        if ctr % chk == 0:
            delta = numpy.linalg.norm(b - a.dot(x), 2)
            if delta < delta_0 * eps:
                break
    return x, ctr


def pde_imp_gs_py(size_h, size_t, init, time, eps, chk):
    """Evolve the PDE using implicit scheme, Gauss-Seidel and Python"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    a_u = -scipy.sparse.triu(a, k=1, format="csr")
    a_dl = scipy.sparse.tril(a.transpose(), k=0, format="csc")
    a_dl_fac = scipy.sparse.linalg.splu(a_dl, permc_spec="NATURAL")
    ctr = 0
    for i in range(m):
        u, temp = solve_gs(a, a_u, a_dl_fac, u, u, eps, chk)
        ctr += temp
    u = u.reshape((n-1, n-1))
    return u, ctr


def solve_sd(mat, vec, init, eps):
    """Steepest descent to solve a system"""
    a, b, x = mat, vec, init
    r = b - a.dot(x)
    rho = r.dot(r)
    rho_0 = rho
    ctr = 0
    while True:
        w = a.dot(r)
        alpha = rho / r.dot(w)
        x = x + alpha * r
        r = r - alpha * w
        rho = r.dot(r)
        ctr += 1
        if rho < rho_0 * eps**2:
            break
    return x, ctr


def pde_imp_sd_py(size_h, size_t, init, time, eps):
    """Evolve the PDE using implicit scheme, steepest descent and Python"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    ctr = 0
    for i in range(m):
        u, temp = solve_sd(a, u, u, eps)
        ctr += temp
    u = u.reshape((n-1, n-1))
    return u, ctr


def solve_cg(mat, vec, init, eps):
    """Conjugate gradient to solve a system"""
    a, b, x = mat, vec, init
    r = b - a.dot(x)
    rho = r.dot(r)
    rho_0 = rho
    ctr = 0
    while True:
        if ctr == 0:
            p = r
        else:
            beta = rho / rho_old
            p = r + beta * p
        w = a.dot(p)
        alpha = rho / p.dot(w)
        x = x + alpha * p
        r = r - alpha * w
        rho_old = rho
        rho = r.dot(r)
        ctr += 1
        if rho < rho_0 * eps**2:
            break
    return x, ctr


def pde_imp_cg_py(size_h, size_t, init, time, eps):
    """Evolve the PDE using implicit scheme, conjugate gradient and Python"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    ctr = 0
    for i in range(m):
        u, temp = solve_cg(a, u, u, eps)
        ctr += temp
    u = u.reshape((n-1, n-1))
    return u, ctr


def step_mg(mats, mat_us, mat_dl_facs, ints, nums, vec, init):
    """Multigrid step"""
    b, x = vec, init
    a, a_u, a_dl_fac = mats[0], mat_us[0], mat_dl_facs[0]
    w = a_dl_fac.solve(b)
    for j in range(nums[0]):
        x = a_dl_fac.solve(a_u.dot(x)) + w
    if len(ints) > 0:
        i = ints[0]
        r = b - a.dot(x)
        b_new = i.transpose().dot(r)
        x_new = numpy.zeros(i.shape[1])
        x_new = step_mg(
            mats[1:], mat_us[1:], mat_dl_facs[1:],
            ints[1:], nums[1:],
            b_new, x_new
        )
        x += i.dot(x_new)
    return x


def solve_mg(mats, mat_us, mat_dl_facs, ints, nums, vec, init, eps, chk):
    """Multigrid to solve a system"""
    b, x = vec, init
    a = mats[0]
    r = b - a.dot(x)
    delta_0 = numpy.linalg.norm(r, 2)
    ctr = 0
    while True:
        x = step_mg(mats, mat_us, mat_dl_facs, ints, nums, b, x)
        ctr += 1
        if ctr % chk == 0:
            r = b - a.dot(x)
            delta = numpy.linalg.norm(r, 2)
            if delta < delta_0 * eps:
                break
    return x, ctr


def pde_imp_mg_py(size_h, size_t, init, time, nums, eps, chk):
    """Evolve the PDE using implicit scheme, multigrid and Python"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    mats, mat_us, mat_dl_facs, ints = [], [], [], []
    temp, j = n, 0
    while True:
        a_u = -scipy.sparse.triu(a, k=1, format="csr")
        a_dl = scipy.sparse.tril(a.transpose(), k=0, format="csc")
        a_dl_fac = scipy.sparse.linalg.splu(a_dl, permc_spec="NATURAL")
        mats.append(a)
        mat_us.append(a_u)
        mat_dl_facs.append(a_dl_fac)
        j += 1
        if j >= len(nums):
            break
        temp //= 2
        i = utils.mat_int(temp)
        ints.append(i)
        a = i.transpose().dot(a.dot(i))
    ctr = 0
    for i in range(m):
        u, temp = solve_mg(mats, mat_us, mat_dl_facs, ints, nums, u, u, eps, chk)
        ctr += temp
    u = u.reshape((n-1, n-1))
    return u, ctr


def pde_imp_cho_c(size_h, size_t, init, time):
    """Evolve the PDE using implicit scheme, Cholesky decomposition and C"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    ctr = exts.pde_imp_cho_wrapper(a.data, a.indices, a.indptr, u, n-1, m)
    u = u.reshape((n-1, n-1))
    return u, ctr


def pde_imp_gs_c(size_h, size_t, init, time, eps, chk):
    """Evolve the PDE using implicit scheme, Gauss-Seidel and C"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    ctr = exts.pde_imp_gs_wrapper(a.data, a.indices, a.indptr, u, eps, chk, m)
    u = u.reshape((n-1, n-1))
    return u, ctr


def pde_imp_sd_c(size_h, size_t, init, time, eps):
    """Evolve the PDE using implicit scheme, steepest descent and C"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    ctr = exts.pde_imp_sd_wrapper(a.data, a.indices, a.indptr, u, eps, m)
    u = u.reshape((n-1, n-1))
    return u, ctr


def pde_imp_cg_c(size_h, size_t, init, time, eps):
    """Evolve the PDE using implicit scheme, conjugate gradient and C"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    ctr = exts.pde_imp_cg_wrapper(a.data, a.indices, a.indptr, u, eps, m)
    u = u.reshape((n-1, n-1))
    return u, ctr


def pde_imp_mg_c(size_h, size_t, init, time, num, eps, chk):
    """Evolve the PDE using implicit scheme, multigrid and C"""
    n, m, u, t = size_h, size_t, init, time
    u = u.flatten()
    num = numpy.array(num, dtype=numpy.int32)
    dx, dt = 1.0 / n, t / m
    lap = utils.mat_lap(n)
    eye = scipy.sparse.eye((n-1)**2).tocsr()
    a = eye + dt / dx**2 * lap
    datas, inds, ptrs = [], [], []
    datas.append(a.data)
    inds.append(a.indices)
    ptrs.append(a.indptr)
    temp = n
    for j in range(len(num) - 1):
        temp //= 2
        i = utils.mat_int(temp)
        datas.append(i.data)
        inds.append(i.indices)
        ptrs.append(i.indptr)
    ctr = exts.pde_imp_mg_wrapper(n, datas, inds, ptrs, u, num, eps, chk, m)
    u = u.reshape((n-1, n-1))
    return u, ctr
